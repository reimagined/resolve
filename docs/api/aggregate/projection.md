---
id: projection
title: Projection
---

:::info TypeScript Support

An aggregate projection object has an associated TypeScript type:

- Type Name - `AggregateProjection`
- Containing Package - `@resolve-js/core`

:::

An aggregate projection is an object of the following structure:

```js
const projection = {
  // The *Init* function creates the initial aggregate state object.
  Init: () => initialState,
  // An event handler function is associated with an event type.
  // It receives the aggregate state and an incoming event
  // and returns the updated state.
  [EVENT_TYPE]: (state, event) => {
    ...
    return newState
  }
  [EVENT_TYPE2]: (state, event) => ...
  [EVENT_TYPE3]: (state, event) => ...
  ...
}
```

An event handler implementation receives the following arguments:

| Argument Name | Description                                                   |
| ------------- | ------------------------------------------------------------- |
| state         | The aggregate state that is an object of arbitrary structure. |
| event         | An [event](../event.md) object.                               |

An event handler should return a new state object.
