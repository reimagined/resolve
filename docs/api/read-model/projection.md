---
id: projection
title: Projection
---

:::info TypeScript Support

A read model projection has an associated TypeScript type:

- Type Name - `ReadModel`
- Package - `@resolve-js/core`

:::

A read model projection is an object of the following structure:

```js
const projection = {
  // The *Init* function initializes the read model's persistent store.
  Init: async (store) => { ... }
  // An event handler function is associated with an event type.
  // It receives the read model store and an incoming event
  // and updates the store based on the event's data.
  [EVENT_TYPE1]: async (store, event, context) -> { ... }
  [EVENT_TYPE2]: async (store, event, context) -> { ... }
}
```

An event handler implementation receives the following arguments:

| Argument Name | Description                                                                                              |
| ------------- | -------------------------------------------------------------------------------------------------------- |
| store         | Exposes [API](../read-model/store.md) used to communicate with the read model's persistent data storage. |
| event         | An object that contains the incoming event's data.                                                       |
| context       | An object that contains data and functions related to the current operation.                             |

## context

The `context` argument is an object with the following fields:

| Field Name | Description                                               |
| ---------- | --------------------------------------------------------- |
| encrypt    | The user-defined [encrypt](../../encryption.md) function. |
| decrypt    | The user-defined [decrypt](../../encryption.md) function. |

This object can also contain additional fields added by [middleware](../../middleware.md).
