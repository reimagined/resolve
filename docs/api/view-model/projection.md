---
id: projection
title: Projection
---

:::info TypeScript Support

A view model projection object has an associated TypeScript type:

- Type Name - `ViewModelProjection`
- Package - `@resolve-js/core`

:::

A view model projection is an object of the following structure:

```js
const projection = {
  // The *Init* function creates the view model's initial state object.
  Init: () => initialState,
  // An event handler function is associated with an event type.
  // It receives the view model's state and an incoming event
  // and returns the updated state.
  [EVENT_TYPE]: (state, event) => {
    ...
    return newState
  }
  [EVENT_TYPE2]: (state, event) => ...
  [EVENT_TYPE3]: (state, event) => ...
  ...
}
```

An event handler implementation receives the following arguments:

| Argument Name | Description                                                                                                                                    |
| ------------- | ---------------------------------------------------------------------------------------------------------------------------------------------- |
| state         | The view model's state that is an object of arbitrary structure.                                                                               |
| event         | An [event](../event.md) object.                                                                                                                |
| args          | Arguments attached to the request.                                                                                                             |
| context       | An object that contains functions and data related to the current operation (see the [Event Handler Context](#event-handler-context) section.) |

An event handler should return a new state object.

## Context

A view model event handler context is an object with the following fields:

| Field Name | Description                                               |
| ---------- | --------------------------------------------------------- |
| jwt        | The JSON Web Token attached to the request.               |
| encrypt    | The user-defined [encrypt](../../encryption.md) function. |
| decrypt    | The user-defined [decrypt](../../encryption.md) function. |
