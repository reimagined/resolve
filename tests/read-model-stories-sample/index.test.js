import givenEvents from 'resolve-testing-tools'
import createReadModelLiteAdapter from 'resolve-readmodel-lite'
// import createReadModelMYSQLAdapter from 'resolve-readmodel-mysql'
// import createReadModelMongoAdapter from 'resolve-readmodel-mongo'

import projection from './projection'
import resolvers from './resolvers'

describe('Read-model sample', () => {
  let adapter = null
  beforeEach(async () => {
    adapter = createReadModelLiteAdapter({
      databaseFile: ':memory:'
    })
    // adapter = createReadModelMYSQLAdapter({
    //   host: 'localhost',
    //   port: 3306,
    //   user: 'root',
    //   password: '',
    //   database: `Stories`
    // })
    // adapter = createReadModelMongoAdapter({
    //   url: 'mongodb://127.0.0.1:27017/Stories'
    // })
    try {
      await adapter.drop(null, 'Stories')
    } catch (e) {}
  })
  afterEach(async () => {
    try {
      await adapter.drop(null, 'Stories')
    } catch (e) {}
    adapter = null
  })

  const events = [
    {
      aggregateId: 'story-id-1',
      type: 'STORY_CREATED',
      payload: 'Text 1.0'
    },
    {
      aggregateId: 'story-id-2',
      type: 'STORY_CREATED',
      payload: 'Text 2.0'
    },
    {
      aggregateId: 'story-id-3',
      type: 'STORY_CREATED',
      payload: 'Text 3.0'
    },
    {
      aggregateId: 'story-id-4',
      type: 'STORY_CREATED',
      payload: 'Text 4.0'
    },
    {
      aggregateId: 'story-id-5',
      type: 'STORY_CREATED',
      payload: 'Text 5.0'
    },
    {
      aggregateId: 'story-id-1',
      type: 'STORY_UPDATED',
      payload: 'Text 1.1'
    },
    {
      aggregateId: 'story-id-2',
      type: 'STORY_UPDATED',
      payload: 'Text 2.1'
    },
    {
      aggregateId: 'story-id-2',
      type: 'STORY_UPDATED',
      payload: 'Text 2.2'
    },
    {
      aggregateId: 'story-id-3',
      type: 'STORY_UPDATED',
      payload: 'Text 3.1'
    },
    {
      aggregateId: 'story-id-3',
      type: 'STORY_UPDATED',
      payload: 'Text 3.2'
    },
    {
      aggregateId: 'story-id-3',
      type: 'STORY_UPDATED',
      payload: 'Text 3.3'
    },
    {
      aggregateId: 'story-id-5',
      type: 'STORY_UPDATED',
      payload: 'Text 5.1'
    },
    {
      aggregateId: 'story-id-5',
      type: 'STORY_UPDATED',
      payload: 'Text 5.2'
    },
    {
      aggregateId: 'story-id-2',
      type: 'STORY_FLAGGED_FOR_DELETION'
    },
    {
      aggregateId: 'story-id-3',
      type: 'STORY_FLAGGED_FOR_DELETION'
    },
    {
      aggregateId: 'story-id-3',
      type: 'STORY_DELETED'
    }
  ]

  test(`resolve "getStoryById" ({ id: 'story-id-1' })`, async () => {
    expect(
      await givenEvents(events)
        .readModel({
          name: 'Stories',
          projection,
          resolvers,
          adapter
        })
        .getStoryById({ id: 'story-id-1' })
    ).toMatchSnapshot(`getStoryById({ id: 'story-id-1' })`)
  })

  test(`resolve "getStoryById" ({ id: 'story-id-2' })`, async () => {
    expect(
      await givenEvents(events)
        .readModel({
          name: 'Stories',
          projection,
          resolvers,
          adapter
        })
        .getStoryById({ id: 'story-id-2' })
    ).toMatchSnapshot(`getStoryById({ id: 'story-id-2' })`)
  })

  test(`resolve "getStoryById" ({ id: 'story-id-3' })`, async () => {
    expect(
      await givenEvents(events)
        .readModel({
          name: 'Stories',
          projection,
          resolvers,
          adapter
        })
        .getStoryById({ id: 'story-id-3' })
    ).toMatchSnapshot(`getStoryById({ id: 'story-id-3' })`)
  })

  test(`resolve "getStoryById" ({ id: 'story-id-4' })`, async () => {
    expect(
      await givenEvents(events)
        .readModel({
          name: 'Stories',
          projection,
          resolvers,
          adapter
        })
        .getStoryById({ id: 'story-id-4' })
    ).toMatchSnapshot(`getStoryById({ id: 'story-id-4' })`)
  })

  test(`resolve "getStoryById" ({ id: 'story-id-5' })`, async () => {
    expect(
      await givenEvents(events)
        .readModel({
          name: 'Stories',
          projection,
          resolvers,
          adapter
        })
        .getStoryById({ id: 'story-id-5' })
    ).toMatchSnapshot(`getStoryById({ id: 'story-id-5' })`)
  })

  test(`resolve "getCountStories" ()`, async () => {
    expect(
      await givenEvents(events)
        .readModel({
          name: 'Stories',
          projection,
          resolvers,
          adapter
        })
        .getCountStories()
    ).toMatchSnapshot(`getCountStories()`)
  })

  test(`resolve "getStoriesByIds" ({ ids: ['story-id-1', 'story-id-2', 'story-id-3', 'story-id-4', 'story-id-5'] })`, async () => {
    expect(
      await givenEvents(events)
        .readModel({
          name: 'Stories',
          projection,
          resolvers,
          adapter
        })
        .getStoriesByIds({
          ids: [
            'story-id-1',
            'story-id-2',
            'story-id-3',
            'story-id-4',
            'story-id-5'
          ]
        })
    ).toMatchSnapshot(
      `.getStoriesByIds({ ids: ['story-id-1', 'story-id-2', 'story-id-3', 'story-id-4', 'story-id-5'] })`
    )
  })

  test(`resolve "getStoriesByPage" ({ skip: 0, limit: 2, ascending: true })`, async () => {
    expect(
      await givenEvents(events)
        .readModel({
          name: 'Stories',
          projection,
          resolvers,
          adapter
        })
        .getStoriesByPage({ skip: 0, limit: 2, ascending: true })
    ).toMatchSnapshot(
      `.getStoriesByPage({ skip: 0, limit: 2, ascending: true })`
    )
  })

  test(`resolve "getStoriesByPage" ({ skip: 2, limit: 2, ascending: true })`, async () => {
    expect(
      await givenEvents(events)
        .readModel({
          name: 'Stories',
          projection,
          resolvers,
          adapter
        })
        .getStoriesByPage({ skip: 2, limit: 2, ascending: true })
    ).toMatchSnapshot(
      `.getStoriesByPage({ skip: 2, limit: 2, ascending: true })`
    )
  })

  test(`resolve "getStoriesByPage" ({ skip: 0, limit: 2, ascending: false })`, async () => {
    expect(
      await givenEvents(events)
        .readModel({
          name: 'Stories',
          projection,
          resolvers,
          adapter
        })
        .getStoriesByPage({ skip: 0, limit: 2, ascending: false })
    ).toMatchSnapshot(
      `.getStoriesByPage({ skip: 0, limit: 2, ascending: false })`
    )
  })

  test(`resolve "getStoriesByPage" ({ skip: 2, limit: 2, ascending: false })`, async () => {
    expect(
      await givenEvents(events)
        .readModel({
          name: 'Stories',
          projection,
          resolvers,
          adapter
        })
        .getStoriesByPage({ skip: 2, limit: 2, ascending: false })
    ).toMatchSnapshot(
      `.getStoriesByPage({ skip: 2, limit: 2, ascending: false })`
    )
  })

  test(`resolve "getStoriesWithRangedVersion" ({ minVersion: 1, maxVersion: 3, openRange: false })`, async () => {
    expect(
      await givenEvents(events)
        .readModel({
          name: 'Stories',
          projection,
          resolvers,
          adapter
        })
        .getStoriesWithRangedVersion({
          minVersion: 1,
          maxVersion: 3,
          openRange: false
        })
    ).toMatchSnapshot(
      `.getStoriesWithRangedVersion({ minVersion: 1, maxVersion: 3, openRange: false })`
    )
  })

  test(`resolve "getStoriesWithRangedVersion" ({ minVersion: 1, maxVersion: 3, openRange: true })`, async () => {
    expect(
      await givenEvents(events)
        .readModel({
          name: 'Stories',
          projection,
          resolvers,
          adapter
        })
        .getStoriesWithRangedVersion({
          minVersion: 1,
          maxVersion: 3,
          openRange: true
        })
    ).toMatchSnapshot(
      `.getStoriesWithRangedVersion({ minVersion: 1, maxVersion: 3, openRange: true })`
    )
  })

  test(`resolve "getStoryVersionById" ({ id: 'story-id-1' })`, async () => {
    expect(
      await givenEvents(events)
        .readModel({
          name: 'Stories',
          projection,
          resolvers,
          adapter
        })
        .getStoryVersionById({ id: 'story-id-1' })
    ).toMatchSnapshot(`.getStoryVersionById({ id: 'story-id-1' })`)
  })
})
