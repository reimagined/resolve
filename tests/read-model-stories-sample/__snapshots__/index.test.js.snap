// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Read-model sample resolve "getCountStories" (): getCountStories() 1`] = `
Object {
  "data": 4,
}
`;

exports[`Read-model sample resolve "getStoriesByIds" ({ ids: ['story-id-1', 'story-id-2', 'story-id-3', 'story-id-4', 'story-id-5'] }): .getStoriesByIds({ ids: ['story-id-1', 'story-id-2', 'story-id-3', 'story-id-4', 'story-id-5'] }) 1`] = `
Object {
  "data": Array [
    Object {
      "active": true,
      "id": "story-id-1",
      "text": "Text 1.1",
      "version": 1,
    },
    Object {
      "active": null,
      "id": "story-id-2",
      "text": "Text 2.2",
      "version": 2,
    },
    Object {
      "active": true,
      "id": "story-id-4",
      "text": "Text 4.0",
      "version": 0,
    },
    Object {
      "active": true,
      "id": "story-id-5",
      "text": "Text 5.2",
      "version": 2,
    },
  ],
}
`;

exports[`Read-model sample resolve "getStoriesByPage" ({ skip: 0, limit: 2, ascending: false }): .getStoriesByPage({ skip: 0, limit: 2, ascending: false }) 1`] = `
Object {
  "data": Array [
    Object {
      "active": true,
      "id": "story-id-5",
      "text": "Text 5.2",
      "version": 2,
    },
    Object {
      "active": true,
      "id": "story-id-4",
      "text": "Text 4.0",
      "version": 0,
    },
  ],
}
`;

exports[`Read-model sample resolve "getStoriesByPage" ({ skip: 0, limit: 2, ascending: true }): .getStoriesByPage({ skip: 0, limit: 2, ascending: true }) 1`] = `
Object {
  "data": Array [
    Object {
      "active": true,
      "id": "story-id-1",
      "text": "Text 1.1",
      "version": 1,
    },
    Object {
      "active": null,
      "id": "story-id-2",
      "text": "Text 2.2",
      "version": 2,
    },
  ],
}
`;

exports[`Read-model sample resolve "getStoriesByPage" ({ skip: 2, limit: 2, ascending: false }): .getStoriesByPage({ skip: 2, limit: 2, ascending: false }) 1`] = `
Object {
  "data": Array [
    Object {
      "active": null,
      "id": "story-id-2",
      "text": "Text 2.2",
      "version": 2,
    },
    Object {
      "active": true,
      "id": "story-id-1",
      "text": "Text 1.1",
      "version": 1,
    },
  ],
}
`;

exports[`Read-model sample resolve "getStoriesByPage" ({ skip: 2, limit: 2, ascending: true }): .getStoriesByPage({ skip: 2, limit: 2, ascending: true }) 1`] = `
Object {
  "data": Array [
    Object {
      "active": true,
      "id": "story-id-4",
      "text": "Text 4.0",
      "version": 0,
    },
    Object {
      "active": true,
      "id": "story-id-5",
      "text": "Text 5.2",
      "version": 2,
    },
  ],
}
`;

exports[`Read-model sample resolve "getStoriesWithRangedVersion" ({ minVersion: 1, maxVersion: 3, openRange: false }): .getStoriesWithRangedVersion({ minVersion: 1, maxVersion: 3, openRange: false }) 1`] = `
Object {
  "data": Array [
    Object {
      "active": null,
      "id": "story-id-2",
      "text": "Text 2.2",
      "version": 2,
    },
    Object {
      "active": true,
      "id": "story-id-5",
      "text": "Text 5.2",
      "version": 2,
    },
  ],
}
`;

exports[`Read-model sample resolve "getStoriesWithRangedVersion" ({ minVersion: 1, maxVersion: 3, openRange: true }): .getStoriesWithRangedVersion({ minVersion: 1, maxVersion: 3, openRange: true }) 1`] = `
Object {
  "data": Array [
    Object {
      "active": true,
      "id": "story-id-1",
      "text": "Text 1.1",
      "version": 1,
    },
    Object {
      "active": null,
      "id": "story-id-2",
      "text": "Text 2.2",
      "version": 2,
    },
    Object {
      "active": true,
      "id": "story-id-5",
      "text": "Text 5.2",
      "version": 2,
    },
  ],
}
`;

exports[`Read-model sample resolve "getStoryById" ({ id: 'story-id-1' }): getStoryById({ id: 'story-id-1' }) 1`] = `
Object {
  "data": Object {
    "active": true,
    "id": "story-id-1",
    "text": "Text 1.1",
    "version": 1,
  },
}
`;

exports[`Read-model sample resolve "getStoryById" ({ id: 'story-id-2' }): getStoryById({ id: 'story-id-2' }) 1`] = `
Object {
  "data": Object {
    "active": null,
    "id": "story-id-2",
    "text": "Text 2.2",
    "version": 2,
  },
}
`;

exports[`Read-model sample resolve "getStoryById" ({ id: 'story-id-3' }): getStoryById({ id: 'story-id-3' }) 1`] = `
Object {
  "data": null,
}
`;

exports[`Read-model sample resolve "getStoryById" ({ id: 'story-id-4' }): getStoryById({ id: 'story-id-4' }) 1`] = `
Object {
  "data": Object {
    "active": true,
    "id": "story-id-4",
    "text": "Text 4.0",
    "version": 0,
  },
}
`;

exports[`Read-model sample resolve "getStoryById" ({ id: 'story-id-5' }): getStoryById({ id: 'story-id-5' }) 1`] = `
Object {
  "data": Object {
    "active": true,
    "id": "story-id-5",
    "text": "Text 5.2",
    "version": 2,
  },
}
`;

exports[`Read-model sample resolve "getStoryVersionById" ({ id: 'story-id-1' }): .getStoryVersionById({ id: 'story-id-1' }) 1`] = `
Object {
  "data": 1,
}
`;
